### ISSUE TITLE
Batoto closing

### ISSUE THREAD

**Issue by Mereck13579** [Deathspike/mangarack#79]
Well https://bato.to/forums/topic/28840-batoto-will-be-closing-permanently/

Now we'll be reliant solely on Mangafox and Kissmanga. Since I'm using a server for all my scrapping my IP has already been preventively banned...
Maybe it's time to find another source(s) to add to Mangarack?

**Comment by Deathspike** [comment]
Aw, hell. I'm all up for other sources if you've got good recommendations. As always, anyone is welcome to take a stab at adding another provider, but to make me do it: convince me why a source recommendation should be added :-)

**Comment by Mereck13579** [comment]
I'd say "let's wait until an alternative arises" but we might not see any and/or it might take quite a bit of time.
Unfortunately, right now, I don't see any good alternative for batoto... except maybe for madokami but that's a really different story.

**Comment by XellossMetallium** [comment]
Yes, since Batoto was intended for higher quality scans nothing comes to mind to replace it. 
Madokami strongly relies on Batoto, so it will be affected too (at least a little), but it could be a good choice to cut to the chase.
Honestly that was unexpected.

**Comment by Deathspike** [comment]
I do feel like this is a good opportunity to look at alternatives. Having just two remaining providers feels odd. I would like to take a look at other sites for redundancy purposes, if anything. But, I guess we'll have to wait and see what happens. I hope someone else begins a scanlation-friendly site. Time will tell :-)

**Comment by XellossMetallium** [comment]
This is a first candidate, born with the purpose to replace Batoto (in fact, they imported Batoto metadata).
https://mangadex.com

Should I find others I'll add them here. I suppose we'll need to monitor the heirs for several weeks.

**Comment by Mereck13579** [comment]
Indeed, it seems like some scanlation groups moved over to mangadex as well so this could be the one.

**Comment by XellossMetallium** [comment]
Bato.to domain was resurrected, but for now it looks like it will take the direction of all the other low-quality providers.

**Comment by Mereck13579** [comment]
Okay we really need something else than Mangafox... now one of my series is present on mangafox but... its chapters are just incomplete. For instance on Please don't bully me, Nagatoro c4:
![image](https://user-images.githubusercontent.com/7541087/35798224-66ed3a3e-0a62-11e8-990f-2754e18bd150.png)
The same chapter on another random scans provider:
![image](https://user-images.githubusercontent.com/7541087/35798561-73a4311e-0a63-11e8-9a11-cdffb3a362bd.png)

**Comment by Deathspike** [comment]
I'm all in favor of alternatives. But, it should be a fairly complete site that at least has a lot of different genres. Suggestions? mangadex is still on my radar but it barely has anything I'm interested in.

**Comment by Mereck13579** [comment]
Oh, yeah, just bear in mind that while I'm very good at finding flaws / criticizing, I'm quite bad at finding alternatives / solutions. 
Just pointed that out since Mangafox was supposed to be "like bato.to but with inferior image quality / scans".

**Comment by XellossMetallium** [comment]
Well, Mangadex is supposed to replace Batoto... literally. The entire archive of Batoto should be uploaded in due time, so content-wise it should do well. Obviously it will be wiser to wait that they actually catch up with Batoto.

The only true "almost complete" high quality provider, to be honest, I think is Madokami. They scraped content from Batoto and grow further with the users contributions. Yet Madokami would need a different approach, since the structure of the site is far from any online reading site.

**Comment by Deathspike** [comment]
Okay. I've also done some checking and reading yesterday, and it looks like most scanlators are going for mangadex. I'll definitely add this as a content provider at some point. I'm also interested in Madokami, it looks like the registration process requires some IRC actions? I haven't done it yet, but can you share some screenshots of the site structure? I don't mind if it's a little odd if the content is really good.

**Comment by Mereck13579** [comment]
Here you go https://imgur.com/a/YKp8c
The main "problem" with Mangarack would probably be the fact that every team can upload there, meaning multiple releases for the same chapter. We had the same problem with bato.to but I think you designed Mangarack to download the first uploaded one.

**Comment by XellossMetallium** [comment]
As @Mereck13579 said the main issue with Madokami is that the main point of the site is to archive a huge quantity of material, with the online reader as an additional feature. 
The folder structure is a bit complex, as you can see in the screenshots. They actually explained how it works, here:

> All series currently under /# - F/, /G - M/, and /N - Z/ will be be moved into
a new 3-part hierarchy based on a 4-character "bucket", i.e., a normalized part
of the series name. The "bucket" is calculated in the following steps:

1. Convert the series name to uppercase.
2. Remove a leading 'THE', 'A', 'AN' if present.
3. Change any non-alpha characters (including numbers) to underscores.
4. Take the first 4 characters.

This bucket is then converted into three nested directories:

1. The first character of the bucket.
2. The first 2 characters of the bucket, or the first character
   if there is only 1 character in the bucket.
3. The full bucket, usually but not necessarily 4 characters long.

For example:

"/N - Z/Nichijou"             becomes "/N/NI/NICH/Nichijou"
"/# - F/A New World Awaits"   becomes "/N/NE/NEW_/A New World Awaits"
"/# - F/An Arranged Marriage" becomes "/A/AR/ARRA/An Arranged Marriage"
"/N - Z/The Voynich Hotel"    becomes "/V/VO/VOYN/The Voynich Hotel"
"/# - F/1 Love 9"             becomes "/_/__/__LO/1 Love 9"
"/# - F/AS"                   becomes "/A/AS/AS/AS"
"/# - F/A+B"                  becomes "/A/A_/A_B/A+B"


As for the registration yes, it happens via IRC. Yet it seems that they experienced some issue after the closure of Batoto, so they disabled the registration procedure for a couple of weeks. I'll check later.

**Comment by Deathspike** [comment]
Thanks for the information @Mereck13579 @XellossMetallium 

> The main "problem" with Mangarack would probably be the fact that every team can upload there, meaning multiple releases for the same chapter. We had the same problem with bato.to but I think you designed Mangarack to download the first uploaded one.

Yeah, and this is a problem if you wish to follow the same release group. But this requirement essentially came from ComicRack compatability. There was no way to differentiate the same number/volume for it to make any kind of sense to ComicRack, so it would freak out. With MangaRack supplying its own reader in the `5.x` branch, that requirement is gone. I'll be making the adjustments to allow a scraper to provide its own file name. This way, a Madokami provider can just provide the file names AS-IS to yield an exact duplicate of whatever is online. Meanwhile, MangaFox (or FoxNet now) can just yield file names as they were in `4.x`.

The biggest problem I see with the "upload whatever you want in whatever way you see fit" scheme is the fact that there is absolutely no way to detect volumes or chapter numbers. I'll have to come up with a clever solution to extract this kind of information. The only real reason for needing this, is that when you reach the end of a chapter in the reader, you want to go to the next chapter... and for that an order is necessary. I'm thinking of doing a monster regex to try and determine this for all non-normalized chapters.. I'll have to ponder this one a bit!

> The folder structure is a bit complex

Alright this makes sense. This is not an issue for MangaRack. We can just let you provide the link to the subfolder where the files are and clone that. The only 'problem' is that there is no search? That's a bit of a bummer. Do you know if a series can be added multiple times (e.g. different spelling, japanese vs english names, etc)? That would be a problem.

**Comment by XellossMetallium** [comment]
> Alright this makes sense. This is not an issue for MangaRack. We can just let you provide the link to the subfolder where the files are and clone that. The only 'problem' is that there is no search? That's a bit of a bummer. Do you know if a series can be added multiple times (e.g. different spelling, japanese vs english names, etc)? That would be a problem.

Search box is present. It works with the entire database, so also novels pop out (usually with "(Novel)" attached). The series are linked with mangaupdates id to get tags and synopsis. Probably they get alternate titles the same way. I tried searching "Legend of legendary heroes" and got the following results:


Path
--
/Manga/D/DE/DENS/Densetsu no Yuusha no Densetsu
/Raws/Densetsu no Yuusha no Densetsu
/Novels/Densetsu no Yuusha no Densetsu
/Novels/Dai Densetsu no Yuusha no Densetsu
/Novels/Densetsu no Yuusha no Densetsu/!Side Stories
/Novels/Toriaezu Densetsu no Yuusha no Densetsu

**Comment by Deathspike** [comment]
Awesome. I'll definitely add this site as a provider. For 5.0 Preview 2, I made these changes:

https://github.com/Deathspike/mangarack/commit/acbd2b336c10e13bcb0c36124c116f62b01abc51
https://github.com/Deathspike/mangarack/commit/f1d18c5e9d88385b001758655b47948b74c3fd6a

Basically allowing sites with vastly different structures to work. Once I can get an account, I'll write some test code. :-)
